Add upcoming event reminders with templates, migration, and tests

andrei-kolesnik-california opened the pull request:
This PR close #6158 - adds email reminders for upcoming events. Emails are sent to users who have marked themselves as attending an event that starts in the next 24 hours. Includes:
   - New notification type: event:reminder
   - Email templates (.mjml, .txt, and precompiled .html)
   - Migration to register the new notification type
   - Test for verifying email delivery content
Manual testing instructions:
   - Create an event starting ~24 hours from now.
   - Mark a user as “going” to the event.
   - Trigger the background job.
   - Confirm that the reminder email is generated and sent.
   - Email subject and body should include event title and formatted start time.
   
🟡 PR Requires review of email templates by the person responsible for external communication and mail content.

andrei-kolesnik-california committed:
Add upcoming event reminders with templates, migration, and tests

aapeliv commented on app/backend/src/couchers/jobs/handlers.py:
```diff
@@ -1055,3 +1058,61 @@ def gen_randomized_coords(user_id, lat, lng):
+                    select(User)
+                    .join(EventOccurrenceAttendee, EventOccurrenceAttendee.user_id == User.id)
+                    .where(EventOccurrenceAttendee.occurrence_id == occurrence.id)
+                    .where(EventOccurrenceAttendee.attendee_status == "going")
```
The two options are going/maybe. I think we can send to both

andrei-kolesnik-california replied:
Agreed, makes sense to include both 'going' and 'maybe' – will update the query accordingly.

aapeliv commented on app/backend/src/couchers/jobs/handlers.py:
```diff
@@ -1055,3 +1058,61 @@ def gen_randomized_coords(user_id, lat, lng):
+    logger.info("Sending event reminder emails")
+
+    with session_scope() as session:
+        now_time = now()
+        lower_bound = now_time + timedelta(hours=23.5)
+        upper_bound = now_time + timedelta(hours=24.5)
```
The background job is not guaranteed to run exactly every 60 minutes (and one cannot in general guarantee that either), so this should use some kind of mechanism to guarantee idempotency.

I suppose the two options are: add a flag to the event, or add a flag to the user occurrence. The flag means keeping track of if we've sent a reminder yet.

andrei-kolesnik-california replied:
You're right — will add a flag to track whether a reminder was already sent.

aapeliv commented on app/backend/src/couchers/jobs/handlers.py:
```diff
@@ -1055,3 +1058,61 @@ def gen_randomized_coords(user_id, lat, lng):
+                    data=notification_data_pb2.EventReminder(
+                        occurrence_id=occurrence.id,
+                        start_time=occurrence.start_time.isoformat(),
+                        event=events_pb2.Event(
+                            event_id=occurrence.id,
+                            slug=occurrence.event.slug,
+                            title=occurrence.event.title,
+                            start_time=Timestamp(seconds=int(occurrence.start_time.timestamp())),
+                        ),
```
Please use `event_to_pb(session, occurrence, context)` here

andrei-kolesnik-california replied:
Will do

aapeliv commented on app/proto/notification_data.proto:
```diff
@@ -178,6 +178,12 @@ message ReferenceReminder {
+message EventReminder {
+  int64 occurrence_id = 1;
+  string start_time = 2;
+  org.couchers.api.events.Event event = 3;
```
I think the event object should have all the info you need

andrei-kolesnik-california replied:
Yes, agreed — no need to duplicate fields

jesseallhands commented on app/backend/templates/v2/event_reminder.txt:
```diff
@@ -0,0 +1,11 @@
+Hi {{ user.name }},
+
+This is a reminder for the upcoming event:
```
```suggestion
This is a reminder that the upcoming event is happening soon:
```

andrei-kolesnik-california replied:
Yes, sounds better — clearer and more natural.

jesseallhands commented on app/backend/templates/v2/event_reminder.txt:
```diff
@@ -0,0 +1,11 @@
+This is a reminder for the upcoming event:
+
+{{ event_title }}
+Starts at {{ start_time }} UTC
```
I don't think anyone except for tech people will understand UTC. Can we show the start time in the event's local time? Could say something like `Starts at {{ start_time }} local time`

andrei-kolesnik-california replied:
Yes, good point. I'll update the text to remove UTC and clarify it's local.

jesseallhands commented on app/backend/templates/v2/event_reminder.txt:
```diff
@@ -0,0 +1,11 @@
+{{ event_title }}
+Starts at {{ start_time }} UTC
+
+Don’t forget to join on time!
```
```suggestion
Don’t forget to attend!
```

andrei-kolesnik-california replied:
Agreed, 'attend' is more general and appropriate.

jesseallhands reviewed (changes requested):
I didn't test the functionality, but assuming it uses the same template as the other emails we should be good. I requested a few wording changes.

I also think we should change the timing from 24 hours before the event to 12 hours before the event.

andrei-kolesnik-california committed:
Applied all suggested changes:
- Send reminders to both 'going' and 'maybe'
- Add idempotency via a flag to user occurrence
- Use event object to avoid redundant fields in proto
- Improved reminder template wording and formatting
- Replaced UTC with localized time

andrei-kolesnik-california committed:
Corrected time of notification

andrei-kolesnik-california committed:
Fix schema mismatch: add server_default=False for reminder_sent to match migration

andrei-kolesnik-california committed:
Merge branch 'develop' into backend/feature/upcoming-event-reminder

aapeliv commented on app/backend/src/couchers/jobs/handlers.py:
```diff
@@ -1055,3 +1058,56 @@ def gen_randomized_coords(user_id, lat, lng):
+    logger.info("Sending event reminder emails")
+
+    with session_scope() as session:
+        now_time = now()
+        lower_bound = now_time + timedelta(hours=11.5)
+        upper_bound = now_time + timedelta(hours=12.5)
+
+        occurrences = (
+            session.execute(
+                select(EventOccurrence)
+                .where(EventOccurrence.start_time >= lower_bound)
+                .where(EventOccurrence.start_time <= upper_bound)
+            )
+            .scalars()
+            .all()
+        )
```
Look for all events happening in the next 24 hours, that way if this bg job doesn't run for some reason for a bit, it will still work (and the `reminder_sent` flag will stop multiple emails.

```python
        occurrences = (
            session.execute(
                select(EventOccurrence)
                .where(EventOccurrence.start_time <= func.now() + EVENT_REMINDER_TIMEDELTA)
                .where(EventOccurrence.start_time >= func.now())
            )
            .scalars()
            .all()
        )
```

bonus points if you put `EVENT_REMINDER_TIMEDELTA` in `constants.py`

andrei-kolesnik-california replied:
I like this approach. Let me follow this path.

aapeliv commented on app/backend/src/couchers/jobs/handlers.py:
```diff
@@ -1055,3 +1058,56 @@ def gen_randomized_coords(user_id, lat, lng):
+                select(User, EventOccurrenceAttendee)
+                .join(EventOccurrenceAttendee, EventOccurrenceAttendee.user_id == User.id)
+                .where(EventOccurrenceAttendee.occurrence_id == occurrence.id)
+                .where(EventOccurrenceAttendee.attendee_status.in_(["going", "maybe"]))
```
Better to use `AttendeeStatus.going` and `AttendeeStatus.maybe` instead of strings. Also these are the only options, so you don't really need this `where`

andrei-kolesnik-california replied:
Sure, I’ll take it out.

aapeliv commented on app/backend/src/couchers/jobs/handlers.py:
```diff
@@ -1055,3 +1058,56 @@ def gen_randomized_coords(user_id, lat, lng):
+
+                attendee.reminder_sent = True
+
+    session.flush()
```
Change this to be a `session.commit()` right after the `attendee.reminder_sent = True`. That way if the bg job gets interrupted, we don't send multiple reminders.

andrei-kolesnik-california replied:
Will commit right after reminder_sent = True to avoid duplicates.

aapeliv commented on app/backend/src/couchers/migrations/versions/14585a4e1868_add_event__reminder_to_.py:
```diff
@@ -0,0 +1,23 @@
+
+
+def upgrade():
+    op.execute("ALTER TYPE notificationtopicaction ADD VALUE IF NOT EXISTS 'event__reminder'")
```
Please merge these into one migration

andrei-kolesnik-california replied:
Good point, I’ll merge the two into a single migration.

aapeliv commented on app/backend/src/couchers/models.py:
```diff
@@ -2165,6 +2165,8 @@ class EventOccurrenceAttendee(Base):
     user = relationship("User")
     occurrence = relationship("EventOccurrence", backref=backref("attendances", lazy="dynamic"))
 
+    reminder_sent = Column(Boolean, nullable=False, default=False, server_default=expression.false())
```
Oooohhh, this is nice!! I didn't know about `expression.false()`

aapeliv commented on app/backend/templates/v2/event_reminder.mjml:
```diff
@@ -0,0 +1,32 @@
+        </mj-text>
+
+        <mj-text font-size="16px" font-family="Helvetica">
+          Starts at {{ start_time }} local time
```
This needs to be `{{ start_time|v2timestamp(user) }}`

andrei-kolesnik-california replied:
Will update to {{ start_time|v2timestamp(user) }}.

aapeliv commented on app/proto/notification_data.proto:
```diff
@@ -178,6 +178,12 @@ message ReferenceReminder {
+message EventReminder {
+  org.couchers.api.events.Event event = 1;
+  org.couchers.api.core.User user = 2;
+  google.protobuf.Timestamp start_time = 3;
```
This is included in `event.start_time`, so no need to add it separately.

andrei-kolesnik-california replied:
You're right, thanks — I’ll remove the extra start_time field.

aapeliv commented on app/backend/src/couchers/notifications/render.py:
```diff
@@ -625,6 +625,26 @@ def render_notification(user, notification) -> RenderedNotification:
                 push_url=event_link,
             )
+        elif notification.action == "reminder":
+            start_time = v2timestamp(data.start_time, user)
```
Use `data.event.start_time`

andrei-kolesnik-california replied:
I’ll update the render as well.

aapeliv reviewed (commented):
The backend logic is almost there! `expression.false()` is also very nifty!

Are you OK if I push a slightly different email template to this branch @andrei-kolesnik-california ?

andrei-kolesnik-california committed:
Refactor event reminder feature per review: merged migrations, removed redundant start_time, updated render, improved attendee selection, added session.commit

andrei-kolesnik-california committed:
Merge Alembic heads 14585a4e1868 and f176b9395ba1 to resolve multiple head revisions

andrei-kolesnik-california committed:
Pass ruff check

andrei-kolesnik-california committed:
Clean up commented code from CreateEvent

andrei-kolesnik-california committed:
Remove unused merge migration 0640368eaeae; set f176b9395ba1 as down_revision for 14585a4e1868

aapeliv committed:
Update event reminder template

aapeliv reviewed (approved):
LGTM

jesseallhands reviewed (approved):
LGTM!
